Cache bazel to speed up build

joestringer opened the pull request:
Previously we have used the '--batch' bazel option to ensure that the
bazel daemon is not running after building cilium. However, if bazel
doesn't continue running, then we don't make use of its build caching
either. Remove the `--batch` option to speed up builds. Also, don't
always clean the binary; that will force rebuild and slow down builds.
    
In local builds, this shaves as much as 20s off build time.

jrajahalme commented on envoy/Makefile:
```diff
@@ -79,24 +98,28 @@ all: clean-bins release
 	$(BAZEL) $(BAZEL_OPTS) build $(BAZEL_BUILD_OPTS) -c dbg //:envoy
 
-$(CHECK_FORMAT): force-non-root
+$(CHECK_FORMAT): $(CILIUM_ENVOY_SOURCES) $(CILIUM_API_SOURCES)
```
This makes sure that the check_format.py is made available, and does not depend on any of our sources so IMO this change should be reverted.

jrajahalme commented on envoy/Makefile:
```diff
@@ -110,20 +133,22 @@ $(ENVOY_API_SOURCES): $(ENVOY_API_PROTO_PATH)
 
-install: force-root
+install:
+	make force-root
```
Is this change for necessity or for uniformity of style?

jrajahalme replied:
'install' is by nature a target that is always run, if asked for?

jrajahalme commented on envoy/Makefile:
```diff
@@ -110,20 +133,22 @@ $(ENVOY_API_SOURCES): $(ENVOY_API_PROTO_PATH)
 endif
 
-bazel-archive: force-non-root tests clean-bins
+bazel-archive: tests clean-bins
```
These targets are not used right now, I think, but still it does not make sense to build bazel caches as root.

joestringer replied:
This one depended on `tests`, which ran the `force-non-root` target, so that's why I removed it here.

rlenglet commented on envoy/Makefile:
```diff
@@ -79,24 +98,28 @@ all: clean-bins release
 
-envoy: force-non-root
+$(ENVOY_BIN) envoy: $(CILIUM_ENVOY_SOURCES) $(CILIUM_API_SOURCES)
+	make force-non-root
```
make -> $(MAKE)

And change every over occurrence below.

jrajahalme commented on envoy/Makefile:
```diff
@@ -142,18 +167,22 @@ veryclean: force
 
-check: $(CHECK_FORMAT) force-non-root
+check: $(CHECK_FORMAT)
+	make force-non-root
```
Is there a specific reason why a phone target should not be a dependency?

rlenglet reviewed (commented):
LGTM but this will need to be rebased on top of https://github.com/cilium/cilium/pull/2781, and merged after that PR.

jrajahalme reviewed (dismissed):
This change will break the envoy version check, which currently requires the git SHA stamped on Envoy binary version to match the Cilium binary git SHA.

joestringer committed:
envoy: Leave bazel running to speed up builds

Previously we have used the '--batch' bazel option to ensure that the
bazel daemon is not running after building cilium. However, if bazel
doesn't continue running, then we don't make use of its build caching
either. Remove the `--batch` option to speed up builds.

In local builds, this shaves as much as 20s off build time.

Signed-off-by: Joe Stringer <joe@covalent.io>

jrajahalme commented on envoy/Makefile:
```diff
@@ -19,7 +19,7 @@ ENVOY_BINS = $(ENVOY_BIN) ./bazel-bin/cilium_integration_test
 CHECK_FORMAT ?= ./bazel-bin/check_format.py.runfiles/envoy/tools/check_format.py
 
 BAZEL ?= bazel
-BAZEL_OPTS ?= --batch
```
Well need to keep the `--batch` option for the PKG_BUILD case.

joestringer replied:
Sure, I'll update for this.

jrajahalme commented on envoy/Makefile:
```diff
@@ -79,7 +79,7 @@ all: clean-bins release
 BAZEL_BUILD_OPTS =
 
-all: clean-bins envoy $(GO_TARGETS)
+all: envoy $(GO_TARGETS)
```
Need to remove the binaries to force a re-link, otherwise the version check will fail.

joestringer replied:
Ah. In my setup, this saves an extra few seconds. Probably if I really want to get those back I'd need to look at introducing an option to avoid the version check. I don't plan on doing that at this stage.

joestringer replied:
@jrajahalme I wonder if something like this could work for local builds?

```
 $ git di
diff --git a/envoy/Makefile b/envoy/Makefile
index 2be0db3a5fe5..f0d94b813ece 100644
--- a/envoy/Makefile
+++ b/envoy/Makefile
@@ -80,7 +80,7 @@ else
 BAZEL_OPTS ?=
 BAZEL_BUILD_OPTS =
 
-all: clean-bins envoy $(GO_TARGETS)
+all: maybe-clean-bins envoy $(GO_TARGETS)
 endif
 
 debug: envoy-debug $(GO_TARGETS)
@@ -132,6 +132,9 @@ bazel-restore: $(BAZEL_ARCHIVE)
        -mkdir $(dir $(BAZEL_CACHE))
        cd $(dir $(BAZEL_CACHE)) && sudo tar xjf $(BAZEL_ARCHIVE) --warning=no-timestamp
 
+maybe-clean-bins: ../.git
+       $(MAKE) clean-bins
+
 # Remove the binaries to get fresh version SHA
 clean-bins: force
        -rm -f $(ENVOY_BINS)
```
